crypto/fipsmodule/sha/sha3.c

```diff
@@ -109,62 +109,39 @@ uint8_t *SHAKE256(const uint8_t *data, const size_t in_len, uint8_t *out, size_t
   return out;
 }
 
-int SHAKE_Init(KECCAK1600_CTX *ctx, size_t block_size) {
-  // The SHAKE block size depends on the security level of the algorithm only
-  // It is independent of the output size
-  ctx->block_size = block_size;
-  return SHA3_Init(ctx, SHAKE_PAD_CHAR, 0);
-}
-
-
-int SHAKE_Final(uint8_t *md, KECCAK1600_CTX *ctx, size_t len) {
-  ctx->md_size = len;
-  return SHA3_Final(md, ctx);
-}
-
-void SHA3_Reset(KECCAK1600_CTX *ctx) {
+// FIPS202 APIs manage internal input/output buffer on top of Keccak1600 API layer
+void FIPS202_Reset(KECCAK1600_CTX *ctx) {
   memset(ctx->A, 0, sizeof(ctx->A));
   ctx->buf_load = 0;
   ctx->padded = 0;
 }
 
-int SHA3_Init(KECCAK1600_CTX *ctx, uint8_t pad, size_t bit_len) {
-  size_t block_size;
-
-  // The block size is computed differently depending on which algorithm
-  // is calling |SHA3_Init|:
-  //   - for SHA3 we compute it by calling SHA3_BLOCKSIZE(bit_len)
-  //     because the block size depends on the digest bit-length,
-  //   - for SHAKE we take the block size from the context.
-  // We use the given padding character to differentiate between SHA3 and SHAKE.
-  if (pad == SHA3_PAD_CHAR) {
-    block_size = SHA3_BLOCKSIZE(bit_len);
-  } else if (pad == SHAKE_PAD_CHAR) {
-    block_size = ctx->block_size;
-  } else {
+int FIPS202_Init(KECCAK1600_CTX *ctx, uint8_t pad, size_t block_size, size_t bit_len) {
+  if (ctx == NULL) {
     return 0;
   }
-  ctx->padded = 0;
-  
-  if (block_size <= sizeof(ctx->buf)) {
-    SHA3_Reset(ctx);
-    ctx->block_size = block_size;
-    ctx->md_size = bit_len / 8;
-    ctx->pad = pad;
-    return 1;
+
+  if (pad != SHA3_PAD_CHAR && 
+      pad != SHAKE_PAD_CHAR) { 
+    return 0;
   }
-  return 0;
+      
+  if (block_size <= sizeof(ctx->buf)) {
+      FIPS202_Reset(ctx);
+      ctx->block_size = block_size;
+      ctx->md_size = bit_len / 8;
+      ctx->pad = pad;
+      return 1;
+    }
+    return 0;
 }
 
-int SHA3_Update(KECCAK1600_CTX *ctx, const void *data, size_t len) {
+int FIPS202_Update(KECCAK1600_CTX *ctx, const void *data, size_t len) {
   uint8_t *data_ptr_copy = (uint8_t *) data;
   size_t block_size = ctx->block_size;
   size_t num, rem;
 
-  if (len == 0) {
-    return 1;
-  }
-
+  // Case |len| equals 0 is checked in SHA3/SHAKE higher level APIs
   // Process intermediate buffer.
   num = ctx->buf_load;
```

Comment:
If FIPS202_Update isn't static we can't know for sure all callers of it will pass in valid arguments so it also needs to check that ctx is not null.